source/common/router/router.cc

```diff
@@ -838,6 +848,27 @@ void Filter::sendNoHealthyUpstreamResponse() {
                              StreamInfo::ResponseCodeDetails::get().NoHealthyUpstream);
 }
 
+void Filter::giveUpRetryAndShadow() {
+  cluster_->trafficStats()->retry_or_shadow_abandoned_.inc();
+  retry_state_.reset();
+  active_shadow_policies_.clear();
+  request_buffer_overflowed_ = true;
+}
+
+void Filter::onUpstreamConnectionEstablished() {
+  ENVOY_LOG(trace, "Got connection with upstream.");
+  wait_for_connect_ = false;
+  if (!request_buffer_overflowed_ &&
```

Comment:
why do we need a new boolean here?  Can't we just check if the retry state still exists?

Reply:
I see the retry state get reset at multiple places. I guess it is more safe to use the already existing bool `request_buffer_overflowed_` which is only modify in `giveUpRetryAndShadow` to be sure we drop retry and shadow when we pass the retry_shadow_buffer_limit_